pallets/vesting-reward/src/lib.rs

```diff
@@ -0,0 +1,295 @@
+// (at your option) any later version.
+
+// Cent-Chain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Cent-Chain.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Reward module for Centrifuge Parachain Campaigns
+//!
+//! ## Overview
+//! The function of this pallet is to provide the centrifuge specific reward functionality for
+//! contributors of the relay chain crowdloan. In order to provide this functionality the pallet
+//! implements the `Reward` Trait from the `Claim` Pallet.
+//! Before any rewards can be provided to contributors the pallet MUST be initialized, so that the
+//! modules account holds the necessary funds to reward.
+//! All rewards are payed in form of a `vested_transfer` with a fixed vesting time. The vesting time
+//! is defined via the modules `Config` trait.
+//!
+//! ## Callable Functions
+//!
+//! - [`set_vesting_start`] : Origin must be admin or root, and allows to set the start of the vesting
+//!    after the initialization.
+//! - [`set_vesting_period`] : Origin must be admin or root, and allows to set the period of the vesting
+//!    after the initialization.
+//! - [`set_conversion_rate`] : Origin must be admin or root, and allows to set the conversion rate
+//!    between relay chain and native balance after the initialization.
+//! - [`set_direct_payout_ratio`] : Origin must be admin or root, and allows to set the ratio between
+//!    vested and direct payout amount after the initialization.
+//!
+//! ## Implementations
+//!
+//! - [`Reward`](crowdloan_claim::reward::Reward) : Rewarding functionality for contributors in a relay
+//!   chain crowdloan.
+
+// Ensure we're `no_std` when compiling for Wasm.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+pub mod weight;
+
+use frame_support::{decl_module, decl_storage, decl_event, decl_error,
+    Parameter,
+    ensure
+};
+use frame_system::{ensure_root, RawOrigin};
+use sp_runtime::traits::{Member, MaybeSerialize, Convert, CheckedDiv};
+use frame_support::sp_runtime::traits::{AtLeast32BitUnsigned, MaybeSerializeDeserialize, Saturating, CheckedSub};
+use frame_support::dispatch::Codec;
+use frame_support::dispatch::fmt::Debug;
+use frame_support::traits::{Get, Currency, ExistenceRequirement::{KeepAlive}, EnsureOrigin};
+use frame_support::dispatch::DispatchResult;
+use crowdloan_claim::reward::Reward;
+use sp_runtime::{ModuleId, traits::{AccountIdConversion, StaticLookup}, Perbill};
+pub use weight::WeightInfo;
+use sp_runtime::traits::Zero;
+
+pub trait Config: frame_system::Config + pallet_vesting::Config {
+    /// This module emits events, and hence, depends on the runtime's definition of event
```

Comment:
I would suggest that we try, as much as we can, when two specific pallets interact together, that the latter being not tightly coupled, especially when interacting with "native" Substrate pallets (as the code base is moving fast). A loosely-coupled regime seems better, using an associated type instead. It is how, for instance, [ChangeMember](https://substrate.dev/rustdocs/v3.0.0/frame_support/traits/trait.ChangeMembers.html) or [Currency](https://substrate.dev/rustdocs/v3.0.0/frame_support/traits/trait.Currency.html) traits work. Moreover, I would suggest that we put all traits in a specific `traits` folder. What is your feeling?

Reply:
not sure I follow what you mean @tankofzion. Do you want to move all the trait definitions to a common module? I think extending the vesting trait here make sense. Am I missing something?

Reply:
> not sure I follow what you mean @tankofzion. Do you want to move all the trait definitions to a common module? I think extending the vesting trait here make sense. Am I missing something?

Hello Ved. Yes, having traits in a module (when traits are rather "common"). In this, you're probably right, it's more an extension, rather than interaction between two pallets. But in the case it's for interacting with two pallets, loosely coupled regime by means of an associated type in the conf seems better.

Reply:
@tankofzion Since this is not bidirectional communication, I would suggest we not go the route of placing the traits under common module. In case of traits depending on each other, then they should belong to same modules. I always felt having common trait module a bit of a code smell :-D.
I'll leave it to @mustermeiszer to make that decision :-D

Reply:
@vedhavyas I was thinking about the way FRAME support organizes [traits](https://github.com/paritytech/substrate/blob/master/frame/support/src/traits.rs). But for tightly related traits, inside a pallet, it make sense to keep them together. Freddy does not need to change anything for now. It was mostly a thought on how we can organize our code base (e.g. that pallets are not defined in the runtime, like radclaims, but in a separate `pallets` folder, so that to be able to play with different runtimes, if necessary). Just thoughts :)

Reply:
We need to tightly couple here in order to call the `vested_transfer` as it is implemented for `T as pallet_vesting::Config`. But I might be missing something here. 

Regarding the second thing: I think I don't understand the advantage here. If I get it right @tankofzion you are referring to creating an associated type, who has the traitbound of the specific pallet you want to interact with? I don't know if this is possible, to be honest. But could indeed be a benefit.

Reply:
> I get it right @tankofzion you are referring to creating an associated type, who has the traitbound of the specific pallet you

Yes. Look at https://substrate.dev/recipes/pallet-coupling.html for instance.

Reply:
> We need to tightly couple here in order to call the `vested_transfer` as it is implemented for `T as pallet_vesting::Config`. But I might be missing something here.

I would say that you're right. My comment was more for pallets interacting together (dispatch extrinsics or call function).

Reply:
> Regarding the second thing: I think I don't understand the advantage of having a module of traits. Personally I like to have them split.

It is more in the vein of what Substrate does, at least for common primitives, where all traits are reexported from a single module. Easier for later imports.